### ISSUE TITLE
AutoPlay.html doesn't allow user to configure leaderboard and raceboard settings

### ISSUE THREAD

**Issue by aksajhfduwafe** [aksajhfduwafe/bugzilla-to-github-issues-test#2610]
| Field | Value |
|---|---|
| **Reporter** | @axeluhl |
| **Assignee** | @fmittag |
| **Product** | Sailing Race Analytics |
| **Component** | Frontend |
| **Version** | unspecified |
| **Hardware** | All |
| **OS** | All |
| **Severity** | normal |
| **Priority** | P1 |
| **CC (unmapped)** | Milan Cerny (`milan.cerny@sap.com`), Simon Pamies (`s.pamies@the-placeholder-company.com`), Stefan Lacher (`stefan.lacher@hs-lu.de`) |
| **Created** | 2015-02-06T07:57:59Z |
| **Original Bug** | [Bug 2610](https://bugzilla.sapsailing.com/bugzilla/show_bug.cgi?id=2610) |

---

The leaderboard always switches back to the default settings. We need the configuration dialog that we also offer in the AdminConsole and should do the same for the map settings. These settings should be remembered and applied each time the controller switches the views.

**Comment by aksajhfduwafe** [comment]
**@axeluhl** commented:

The implementation approach should use a new "Perspective" concept that allows clients to find out the set of Components used, together with their Settings, and then pull their SettingsDialogComponents into the AutoPlay configuration page.

**Comment by aksajhfduwafe** [comment]
**@fmittag** commented:

Axel, please review... branch bug2610_lifecycle

**Comment by aksajhfduwafe** [comment]
**@axeluhl** commented:

ComponentLifecycle uses a type argument for the component type:

	public interface ComponentLifecycle<C extends Component<S>, S extends Settings, SDC extends SettingsDialogComponent<S>> {
	
Why? The type parameter "C" is not used anywhere in the interface declaration. Is this for "documentation" purposes so that an implementation ties together the component type, its settings type as well as the settings dialog component?

ComponentLifecycleAndSettings follows a slightly different naming convention than ComponentAndSettingsPair. Should we rename "ComponentAndSettingsPair" into "ComponentAndSettings"? d5f4195c91daf5dd3ec4b8b021b3a265d285ab60

The ComponentLifecycleAndSettings.setSettings method is not used anywhere. I'll make the field final and remove the setter. 5453257e0f2441c3c09fb84b6dd2563ddf0df3fd

ComponentLifecycleSettings contains a commented setter method. Removed. 28ddc065f8b9edb9d947adffda7d13a8d68523a3

CompositeLifecycleSettings uses an Iterable that must be searched for settings belonging to a ComponentLifecycle object. Wouldn't a Map be a better implementation choice? The map could be filled upon construction. I see, however, that the use of generics becomes a lot more difficult for the individual ComponentLifecycle/Settings pairs if a map is used as it is not easily possible to specify that an entry's settings type has to be the same as the ComponentLifecycle's. I guess it's OK to iterate through these up to three or four elements for now. Should it ever become a performance issue then we can still consider introducing a redundant map in CompositeLifecycleSettings to accelerate the lookup method.

Why is CompositeTabbedSettingsComponent.getSettings() returning null? shouldn't it return a CompositeSettings object that is assembled from the components? Then again, CompositeTabbedSettingsComponent doesn't seem to be used anywhere (anymore?). Instead, PerspectiveCompositeTabbedSettingsComponent seems to have taken its place. However, PerspectiveCompositeTabbedSettingsComponent.getSettings() also returns null. Why? Shouldn't it properly fill the PerspectiveCompositeSettings object? Then again, PerspectiveCompositeTabbedSettingsDialog is not used anywhere either, and neither is PerspectiveCompositeTabbedSettingsComponent. I removed both classes in 413e640f1bd255fc3a11332f15297dd5c9a0b4c0. Objections? I also removed CompositeTabbedSettingsDialog and CompositeTabbedSettingsComponent in 14daa8f94b3376ef25ca3be7f07cd460773c6104.

LifecycleSettingsDialog was unused. I removed it in commit ebe0077983d0929fc2e7783aa7274f9eb0865c6e.

I added i18n for the perspective names (2d424f005b0a0d3d32498fe5fa1b77ec6a26dd63). I think PerspectiveLifecycle.getPerspectiveName() is redundant to the inherited getLocalizedShortName() which becomes obvious by a delegation to that method being the default implementation in AbstractPerspectiveLifecycle. Furthermore, getPerspectiveName() has no indication as to whether the result is localized or not. I remove the method and replace it by getLocalizedShortName(). 000f6b10b0619e0cbea5b88efaf18ecfac3196f6

AbstractPerspectiveComposite could hold the PerspectiveLifecycle field and do the delegation that is now redundantly implemented in both subclasses existing so far. I currently don't see a reason why this couldn't move to the AbstractPerspectiveComposite class. 20387e32a4df272524fd6a3dc48b3ef476d48910

AbstractPerspectiveWithoutSettings was not used. Removed with commit a56531d6c843e89d3009c8ec97f1929c8cb48d35. AbstractPerspective was unused. Gone in 55a4cb608f74dc6a701e932fefbc7ac6c068960b.

RaceBoardPanel.getComponents() was now redundant with AbstractPerspectiveComposite.getComponents(). Removed in 72f63c72e1bd771758366d69b39c5000d85b24fc.

Why does Perspective need to offer the

    CompositeSettings getSettingsOfComponents();
    void updateSettingsOfComponents(CompositeSettings settingsOfComponents);

methods? Implementing the "Composite" pattern, a Perspective in particular is a Component that has Settings. We discussed that it must be possible to nest perspectives. Looking at the two existing perspectives (LeaderboardWithHeaderPerspective and RaceBoardPanel), I don't see this implemented. I neither see implemented that Autoplay itself is a (nested) perspective which could be configured using the same techniques. I would have expected the SettingsType of a Perspective to be constrained to CompositeSettings, one element of which being for the perspective itself (what is currently returned from getSettings() on the Perspective implementations) and the other elements being for the components presented by the perspective. Only this way does a Perspective object correctly implement the composite pattern, I think. We need to discuss. Stopping the review at this point for now.
